src/cluster.c

```diff
@@ -5922,44 +6115,129 @@ NULL
+             *    n NODE B and returns control back to client C
+             * 4. The following steps can happen in parallel
+             * a. Client C issues SETSLOT n NODE B against node A
+             * b. node B gossips its new slot ownership to the cluster
+             *    including A, A', etc
+             *
+             * Where A is the source primary and B is the destination primary.
+             *
+             * This enforcement is implemented via a new phase called `REPLICAONLY'`
+             * in the setslot subcommand. */
+
+            n = clusterLookupNode(c->argv[4]->ptr, sdslen(c->argv[4]->ptr));
+
+            if (!n) {
+                addReplyErrorFormat(c,"Unknown node %s", (char*)c->argv[4]->ptr);
+                return;
+            }
+            if (nodeIsSlave(n)) {
+                addReplyError(c,"Target node is not a master");
+                return;
+            }
+            /* If this hash slot was served by 'myself' before to switch
+             * make sure there are no longer local keys for this hash slot. */
+            if (server.cluster->slots[slot] == myself && n != myself) {
+                if (countKeysInSlot(slot) != 0) {
+                    addReplyErrorFormat(c,
+                        "Can't assign hashslot %d to a different node "
+                        "while I still hold keys for this hash slot.", slot);
+                    return;
+                }
+            }
+
+            /* Replicate the command to the replicas if there are any */
+            if ((server.cluster->importing_slots_from[slot] != NULL) &&
+                (nodeIsMaster(myself) != 0) &&
+                (myself->numslaves != 0))
+            {
+                /* Remove the last "REPLICAONLY" token so the command
+                 * can be applied as the real "SETSLOT" command on the
+                 * replicas. */
+                serverAssert(c->argc == 6);
+                c->argc--;
+                decrRefCount(c->argv[5]);
+                c->argv[5] = NULL;
+
+                /* We are a primary and this is the first time we see this "setslot"
+                 * command. Force-replicate the setslot command to all of our replicas
+                 * first and only on success will we handle the command.
+                 * Note that
+                 *  1) The repl offset target is set to the master's current repl
+                 *     offset + 1. There is no concern of partial replication because
+                 *     replicas always ack the repl offset at the command boundary.
+                 *  2) This request is idempotent. If the replication cannot be completed
+                 *     in 5000 ms, the user request will be failed. Admins can retry the
+                 *     request safely until it succeeds. */
+                blockForReplication(c, mstime()+5000, server.master_repl_offset+1, myself->numslaves);
+                replicationRequestAckFromSlaves();
```

Comment:
Isn't this still (semi-)synchronous replication?

I thought the idea was to make SETSLOT REPLICAONLY replicate without blocking and just return the number of replicas. Next, the client sends a WAIT. In this way, we don't need to introduce another WAIT-like command that blocks for replication. I think that's what the core team suggested. Are there any issues with this approach?

Reply:
Thanks @zuiderkwast! Yeah I think I missed that "full async" part - I have updated the logic now to go full async. The wait is now initiated by the client now.